### PR TITLE
Added support for impersonation on SQL Server databases using the pym…

### PR DESCRIPTION
**Pull request by tjmateus**
…ssql adapter.

### CATEGORY

Choose one

- [ ] Bug Fix
- [x] Enhancement (new features, refinement)
- [ ] Refactor
- [ ] Add tests
- [ ] Build / Development Environment
- [ ] Documentation

### SUMMARY

Issue #9404 

By default, the impersonation on Superset for SQL Server connections is done by replacing the user of the connection string by the logged in user name. This was not the behaviour we were expecting at DefinedCrowd. We want to have an application user capable of impersonating the logged in user.

The solution I'm proposing is to use the mssql+pymssql SQLAlchemy adapter you already support and use the conn_properties query parameter, which allows us to send SQL queries to the server upon the connection establishment. This query would be EXECUTE AS USER = '<USER_TO_IMPERSONATE>', which will run all the following queries as the logged in user.

### TEST PLAN

####Exploratory test

1. Create a database user for your logged in user. They **must** share the same username;

2. Create an application user on a SQL Server database. This user must be granted
with the `IMPERSONATE` permission as below:

```
GRANT IMPERSONATE ON USER::[<USER_TO_IMPERSONATE>] TO <APPLICATION_USER>;
```

3. On Superset UI, create a new database source with a connection with the following syntax:

```
mssql+pymssql://<username>:<password>@<server>:<port>/<database_name>
```

4. Enable the `Impersonate the logged on user` flag;

5. Run the query `SELECT CURRENT_USER as username;` on SQL Lab;

6. The expected output should be your logged in user and not the application one.

####Unit test

If you think this feature would be useful for community, I will invest time on this kind of tests.

### CODE REVIEW

File: superset/db_engine_specs/mssql.py [hunk 1]
```diff
@@ -76,3 +77,16 @@ def get_sqla_column_type(cls, type_: str) -> Optional[TypeEngine]:
...
+        cls, url: URL, impersonate_user: bool, username: Optional[str]
+    ) -> None:
+        """
+        Modify the SQL Alchemy URL object with the user to impersonate if applicable.
+        :param url: SQLAlchemy URL object
+        :param impersonate_user: Flag indicating if impersonation is enabled
+        :param username: Effective username
+        """
+        if impersonate_user and username is not None:
+            url.query["conn_properties"] = "EXECUTE AS USER = '{}'".format(username)
```
Comment (villebro) [inline superset/db_engine_specs/mssql.py]: We tend to use f-strings, e.g. `abc = f"variable goes {here}"`.

**Review by villebro — COMMENTED** [review]
I assume this is limited to `pymssql`? Given that `pymssql` is deprecated and `pyodbc` to my knowledge is the best practice of connecting to MSSQL these days, we should perhaps add logic to check that the appropriate syntax is being used for the chosen driver. Or alternatively raising an exception if not supported on other connectors.

File: superset/db_engine_specs/mssql.py (hunk 1 above)
Comment (tjmateus) [reply superset/db_engine_specs/mssql.py]: You are right, `pymssql` is deprecated and this feature is limited to that adapter. I've analyzed first `pyodbc` but I didn't find a way to send SQL queries to the server upon the connection establishment. I added logic to check the driver being used and applied this logic only for `pymssql`. For other adapters I just don't know if we should raise an exception or just add the default behaviour, which is replacing the connection string username by the logged in user's. What do you think?

Comment (tjmateus) [reply superset/db_engine_specs/mssql.py]: Hi @villebro, did you have a chance to think on my question? Thanks!

File: superset/db_engine_specs/mssql.py [hunk 2]
```diff
@@ -76,3 +78,19 @@ def get_sqla_column_type(cls, type_: str) -> Optional[TypeEngine]:
...
+        Modify the SQL Alchemy URL object with the user to impersonate if applicable.
+        :param url: SQLAlchemy URL object
+        :param impersonate_user: Flag indicating if impersonation is enabled
+        :param username: Effective username
+        """
+        if impersonate_user and username is not None:
+            if url.get_dialect().driver == MSDialect_pymssql.driver:
+                url.query["conn_properties"] = f"EXECUTE AS USER = '{username}'"
+            else:
+                url.username = username
```
Comment (villebro) [inline superset/db_engine_specs/mssql.py:96]: I would turn this around, ie. check if the driver is `MSDialect_pymssql`, then do the different logic, else call the original method `super(). modify_url_for_impersonation(url, impersonate_user, username)`.

**Review by villebro — CHANGES_REQUESTED** [review]
Looks mostly good, one small change request

File: superset/db_engine_specs/mssql.py (hunk 2 above)
Comment (tjmateus) [reply superset/db_engine_specs/mssql.py]: Makes sense! Already made the requested changes.

**Comment by villebro** [comment]
@tjmateus one more thing, could you add a note under `docs/installation.rst` and add instructions for how this works with `pymssql`? This make it more accessible to other users, too.

**Review by villebro — CHANGES_REQUESTED** [review]
Can you move this to the section above: https://github.com/apache/incubator-superset/blob/master/docs/installation.rst#database-dependencies . There's an optional section below the table for databases with noteworthy comments, I think this would sit well there under a `SQL Server` sub section.

**Review by villebro — APPROVED** [review]
Looking good @tjmateus 👍 Will leave this here for a while for others to chime in, but this looks ready to me.

**Comment by tjmateus** [comment]
> Looking good @tjmateus 👍 Will leave this here for a while for others to chime in, but this looks ready to me.

Thanks for all your inputs, @villebro. They were really useful!

**Pull request closed without merging**